webrtc/README.md

```diff
@@ -0,0 +1,412 @@
+bytes, encoded as an unsigned variable length integer as defined by the
+[multiformats unsigned-varint spec][uvarint-spec].
+
+It is an adaptation from the [QUIC RFC]. When in doubt on the semantics of
+these messages, consult the [QUIC RFC].
+
+``` proto
+syntax = "proto2";
+
+package webrtc.pb;
+
+message Message {
+  enum Flag {
+    // The sender will no longer send messages on the stream.
+    FIN = 0;
+    // The sender will no longer read messages on the stream. Incoming data is
+    // being discarded on receipt.
+    STOP_SENDING = 1;
+    // The sender abruptly terminates the sending part of the stream. The
+    // receiver can discard any data that it already received on that stream.
+    RESET_STREAM = 2;
+  }
+
+  optional Flag flag=1;
+
+  optional bytes message = 2;
+}
+```
+
+Note that "a STOP_SENDING frame requests that the receiving endpoint send a
+RESET_STREAM frame.". See [QUIC RFC - 3.5 Solicited State
+Transitions](https://www.rfc-editor.org/rfc/rfc9000.html#section-3.5).
+
+Encoded messages including their length prefix MUST NOT exceed 16kiB to support
+all major browsers. See ["Understanding message size
+limits"](https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API/Using_data_channels#understanding_message_size_limits).
+Implementations MAY choose to send smaller messages, e.g. to reduce delays
+sending _flagged_ messages.
+
+### Ordering
+
+Implementations MAY expose an unordered byte stream abstraction to the user by
+overriding the default value of `ordered` `true` to `false` when creating a new
+data channel via
+[`RTCPeerConnection.createDataChannel`](https://www.w3.org/TR/webrtc/#dom-peerconnection-createdatachannel).
+
+## Connection Security
+
+Note that the below uses the message framing described in
+[multiplexing](#multiplexing).
+
+While WebRTC offers confidentiality and integrity via TLS, one still needs to
+authenticate the remote peer by its libp2p identity.
+
+After [Connection Establishment](#connection-establishment):
+
+1. _A_ opens a WebRTC datachannel.
+
+2. _A_ starts a Noise `XX` handshake using _A_'s and _B_'s libp2p identity. See
+   [noise-libp2p](https://github.com/libp2p/specs/tree/master/noise).
+
+   Instead of exchanging the TLS certificate fingerprints on the established
+   Noise channel once the Noise handshake succeeded, _A_ and _B_ use the [Noise
+   Prologue](https://noiseprotocol.org/noise.html#prologue) mechanism, thus
+   saving one round trip.
+
+   More specifically _A_ and _B_ set the Noise _Prologue_ to
+   `libp2p-webrtc-noise:<FINGERPRINTS>` before starting the actual Noise
+   handshake. `<FINGERPRINTS>` is the concatenation of the of the two TLS
+   fingerprints of _A_ and _B_ in their multihash byte representation, sorted in
+   ascending order.
+
+3. On success of the authentication handshake, the used datachannel is
+   closed and the plain WebRTC connection is used with its multiplexing
+   capabilities via datachannels. See [Multiplexing](#multiplexing).
```

Comment:
I really don't know if that Noise thing is secure.
The objective of the Noise handshake, normally, is to derive a common secret. If the remote lies and sends a key whose private key is unknown to it, it won't be able to decrypt the communication afterwards.
But here, we just throw the common secret away afterwards.

What prevents a peer A from simply stealing the Noise static key from peer B? Sure, A doesn't know the private key associated with B's Noise static key. But, correct me if I'm wrong, the Noise handshake doesn't actually attempt to check whether the remote possesses its the private key of its Noise static key. Instead, it just goes with the idea that the communication won't be decryptable afterwards.

Reply:
> The objective of the Noise handshake, normally, is to derive a common secret.

This isn't totally correct. Depending on the handshake pattern we can also provide mutual authentication. Here we are using the `XX` pattern handshake:

```
XX:
  -> e
  <- e, ee, s, es
  -> s, se
```

The first message from the responder (second message overall) will contain the authentication data as part of the payload. This includes the associated data with the current value of the handshake hash `h` (which also includes the prologue). When the initiator decrypts this payload it will know that that the responder has access to the private key for `s` because the payload was encrypted with the results of the `ee` and `es` DH outputs.

From the Noise Protocol Framework spec:
> The responder can send its static public key (under encryption) and authenticate itself via a slightly different pattern:
> 
>   -> e
>   <- e, ee, s, es
> 
> In this case, the final ck and k values are a hash of both DH results. Since the es token indicates a DH between the initiator's ephemeral key and the responder's static key, successful decryption by the initiator of the second message's payload serves to authenticate the responder to the initiator.
> 
> Note that the second message's payload may contain a zero-length plaintext, but the payload ciphertext will still contain authentication data (such as an authentication tag or "synthetic IV"), since encryption is with an AEAD mode. The second message's payload can also be used to deliver certificates for the responder's static public key.

Using the prologue here also seems fine since the Noise Protocol framework spec says:
> Noise protocols have a prologue input which allows arbitrary data to be hashed into the h variable. If both parties do not provide identical prologue data, the handshake will fail due to a decryption error. This is useful when the parties engaged in negotiation prior to the handshake and want to ensure they share identical views of that negotiation.

Which seems in line with what we are doing here. Both nodes should agree on what the TLS cert fingerprints were.

Reply:
Resolving here. Thanks @MarcoPolo for the elaborate response.